modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/Processor.java

```diff
@@ -550,86 +550,117 @@ private boolean isObjectClass(TypeMirror type) {
-            checkSuperclassContainAnyAnnotation(clazz, superClazz, PolymorphicConfig.class);
-            
+
+            checkSuperclassContainAnyAnnotation(clazz, superClazz, ConfigurationRoot.class, Config.class);
+
             checkNoConflictFieldNames(clazz, superClazz, fields, fields(superClazz));
-        } else if (clazz.getAnnotation(ConfigurationRoot.class) != null) {
-            checkNotContainsPolymorphicIdField(clazz, ConfigurationRoot.class, fields);
-        } else if (clazz.getAnnotation(Config.class) != null) {
-            checkNotContainsPolymorphicIdField(clazz, Config.class, fields);
         }
     }
-    
+
+    private void validatePolymorphicConfig(TypeElement clazz, List<VariableElement> fields) {
+        checkIncompatibleClassAnnotations(
+                clazz,
+                PolymorphicConfig.class,
+                ConfigurationRoot.class, Config.class, PolymorphicConfigInstance.class
+        );
+
+        checkNotExistSuperClass(clazz, PolymorphicConfig.class);
+
+        List<VariableElement> typeIdFields = collectAnnotatedFields(fields, PolymorphicId.class);
+
+        if (typeIdFields.size() != 1 || fields.indexOf(typeIdFields.get(0)) != 0) {
+            throw new ProcessorException(String.format(
+                    "Class with %s must contain one field with %s and it should be the first in the schema: %s",
+                    simpleName(PolymorphicConfig.class),
+                    simpleName(PolymorphicId.class),
+                    clazz.getQualifiedName()
+            ));
+        }
+    }
+
+    private void validatePolymorphicConfigInstance(TypeElement clazz, List<VariableElement> fields) {
+        checkIncompatibleClassAnnotations(
+                clazz,
+                PolymorphicConfigInstance.class,
+                ConfigurationRoot.class, Config.class
+        );
+
+        checkNotContainsPolymorphicIdField(clazz, PolymorphicConfigInstance.class, fields);
+
+        String id = clazz.getAnnotation(PolymorphicConfigInstance.class).value();
+
+        if (id == null || id.isBlank()) {
+            throw new ProcessorException(String.format(
+                    EMPTY_FIELD_ERROR_FORMAT,
+                    simpleName(PolymorphicConfigInstance.class) + ".id()",
+                    clazz.getQualifiedName()
+            ));
+        }
+
+        checkExistSuperClass(clazz, PolymorphicConfigInstance.class);
+
+        TypeElement superClazz = superClass(clazz);
+
+        checkSuperclassContainAnyAnnotation(clazz, superClazz, PolymorphicConfig.class);
+
+        checkNoConflictFieldNames(clazz, superClazz, fields, fields(superClazz));
+    }
+
+    private void validateConfigurationRoot(TypeElement clazz, List<VariableElement> fields) {
+        checkNotContainsPolymorphicIdField(clazz, ConfigurationRoot.class, fields);
+    }
+
+    private void validateConfig(TypeElement clazz, List<VariableElement> fields) {
+        checkNotContainsPolymorphicIdField(clazz, Config.class, fields);
```

Comment:
`validateConfig()` is a higher-level method than `checkNotContainsPolymorphicIdField()`.

The idea is that for each 'thing' that can be consumed by this processor, there is a separate validation method. A method to validate @Config-annotated, a method to validate @ConfigurationRoot-annotated, and so on.

Look at the current `validate()` code: it's clear that first it validates something common, then it validates every special case, AND it can be clearly seen what this special case is for. On the other hand, if we just inline this method, it would look like

```
        } else if (clazz.getAnnotation(ConfigurationRoot.class) != null) {
            validateConfigurationRoot(clazz, fields);
        } else if (clazz.getAnnotation(Config.class) != null) {
            checkNotContainsPolymorphicIdField(clazz, Config.class, fields);
        }
```

So in the last branch we see technical details of how validation is made, but it's a bit more difficult to understand WHAT specifically is validated there.

So I think the method has value and should not be inlined.

Reply:
So far, only one check, when more is needed, we will make separate methods.

Reply:
The point is not about the number of checks. It is rather about logical layering and ease of reading of `validate()` methods. It brings consistency if any branch related to validation of 'something' is just an invocation of a method called `validateSomething()`, isn't it? :)